posthog/tasks/alerts/trends.py

```diff
@@ -0,0 +1,216 @@
+from typing import TypedDict, NotRequired
+from posthog.tasks.alerts.utils import (
+    AlertEvaluationResult,
+    NON_TIME_SERIES_DISPLAY_TYPES,
+)
+
+
+# TODO: move the TrendResult UI type to schema.ts and use that instead
+class TrendResult(TypedDict):
+    action: dict
+    actions: list[dict]
+    count: int
+    data: list[float]
+    days: list[str]
+    dates: list[str]
+    label: str
+    labels: list[str]
+    breakdown_value: str | int | list[str]
+    aggregated_value: NotRequired[float]
+    status: str | None
+    compare_label: str | None
+    compare: bool
+    persons_urls: list[dict]
+    persons: dict
+    filter: dict
+
+
+def check_trends_alert(alert: AlertConfiguration, insight: Insight, query: TrendsQuery) -> AlertEvaluationResult:
+    if "type" in alert.config and alert.config["type"] == "TrendsAlertConfig":
+        config = TrendsAlertConfig.model_validate(alert.config)
+    else:
+        ValueError(f"Unsupported alert config type: {alert.config}")
+
+    condition = AlertCondition.model_validate(alert.condition)
+    threshold = InsightThreshold.model_validate(alert.threshold.configuration) if alert.threshold else None
+
+    if not threshold:
+        return AlertEvaluationResult(value=0, breaches=[])
+
+    match condition.type:
+        case AlertConditionType.ABSOLUTE_VALUE:
+            if threshold.type != InsightThresholdType.ABSOLUTE:
+                raise ValueError(f"Absolute threshold not configured for alert condition ABSOLUTE_VALUE")
+
+            # want value for current interval (last hour, last day, last week, last month)
+            # depending on the alert calculation interval
+            if _is_non_time_series_trend(query):
+                filters_override = _date_range_override_for_intervals(query)
+            else:
+                # for non time series, it's an aggregated value for full interval
+                # so we need to compute full insight
+                filters_override = None
+
+            calculation_result = calculate_for_query_based_insight(
+                insight,
+                execution_mode=ExecutionMode.RECENT_CACHE_CALCULATE_BLOCKING_IF_STALE,
+                user=None,
+                filters_override=filters_override,
+            )
+
+            if not calculation_result.result:
+                raise RuntimeError(f"No results found for insight with alert id = {alert.id}")
+
+            current_interval_value = _pick_interval_value_from_trend_result(config, query, calculation_result)
+            breaches = _validate_bounds(threshold.bounds, current_interval_value)
+
+            return AlertEvaluationResult(value=current_interval_value, breaches=breaches)
+
+        case AlertConditionType.RELATIVE_INCREASE:
+            if _is_non_time_series_trend(query):
+                raise ValueError(f"Relative alerts not supported for non time series trends")
+
+            # to measure relative increase, we can't alert until current interval has completed
```

Comment:
Because of the limitation, these will only fire once per interval window. Maybe we should stop them from running alerts at an interval that doesn't make sense (setting hourly but this could ever only update weekly)

Also, we can tell them relative increase greater than or relative decrease less than before this point, right?

the interaction between alert run time and these intervals seems complex

Reply:
> Also, we can tell them relative increase greater than or relative decrease less than before this point, right?

Yes we can, that's how I implemented them initially but then got complex as for increase less than/decrease more than, will be checked on different interval. The users asking it only wanted it when the interval completes so I went with this approach as start. Later will look into a way to simplify this more and also alert earlier for other scenarios.